### PR TITLE
Add Maybe to the end bound of proofs (Part 1)

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
There have been several bugs related to the nil key being confused for no end bound.  By converting the end bounds to Maybe[T], it will be more explicit in the code which bound is meant (nil key vs no bound)

### CODE REVIEW

File: x/merkledb/db.go [hunk 1]
```diff
@@ -997,21 +997,22 @@ func (db *merkleDB) VerifyChangeProof(
 	smallestPath := newPath(start)
 
 	// Make sure the start proof, if given, is well-formed.
-	if err := verifyProofPath(proof.StartProof, smallestPath); err != nil {
+	if err := verifyProofPath(proof.StartProof, Some(smallestPath)); err != nil {
```
Comment (dboehm-avalabs) [inline x/merkledb/db.go:1000]: start is always meaningful since a nil start key behaves identically to no start key

File: x/sync/network_server.go [hunk 2]
```diff
@@ -138,11 +138,14 @@ func (s *NetworkServer) HandleChangeProofRequest(
 	requestID uint32,
 	req *pb.SyncGetChangeProofRequest,
 ) error {
+	if req.EndKey == nil {
```
Comment (darioush) [inline x/sync/network_server.go:141]: not sure if we have valid representation with `req.EndKey == nil`, why do we need `IsNothing`?

File: x/merkledb/proof_test.go [hunk 3]
```diff
@@ -845,7 +845,7 @@ func Test_ChangeProof_Syntactic_Verify(t *testing.T) {
 				KeyChanges:        []KeyChange{{Key: []byte{1}, Value: Some([]byte{1})}},
 			},
 			start:       []byte{0},
-			end:         nil, // Also tests start can be after end if end is nil
+			end:         Nothing[[]byte](), // Also tests start can be after end if end is nil
```
Comment (darioush) [inline x/merkledb/proof_test.go:848]: should this comment be updated from nil -> Nothing?

File: x/merkledb/proof.go [hunk 4]
```diff
@@ -693,9 +690,12 @@ func verifyKeyValues(kvs []KeyValue, start, end []byte) error {
 //   - Each key in [proof] is a strict prefix of [keyBytes], except possibly the last.
 //   - If the last element in [proof] is [keyBytes], this is an inclusion proof.
 //     Otherwise, this is an exclusion proof and [keyBytes] must not be in [proof].
-func verifyProofPath(proof []ProofNode, keyPath path) error {
-	provenKey := keyPath.Serialize()
+func verifyProofPath(proof []ProofNode, keyPath Maybe[path]) error {
```
Comment (darioush) [inline x/merkledb/proof.go:693]: could we update the comment here? (should refer to keyPath instead of keyBytes)

Comment (darioush) [reply x/merkledb/proof.go:693]: what is the expected behavior if `keyPath` is Nothing? It seems we try to serialize the value regardless?

File: x/merkledb/proof.go [hunk 5]
```diff
@@ -671,15 +669,14 @@ func verifyKeyChanges(kvs []KeyChange, start, end []byte) error {
 // If [start] is nil, there is no lower bound on acceptable keys.
 // If [end] is nil, there is no upper bound on acceptable keys.
 // If [kvs] is empty, returns nil.
-func verifyKeyValues(kvs []KeyValue, start, end []byte) error {
+func verifyKeyValues(kvs []KeyValue, start []byte, end Maybe[[]byte]) error {
```
Comment (darioush) [inline x/merkledb/proof.go:672]: could we update the comment for this fn to refer to Nothing instead of nil as appropriate?

File: x/merkledb/proof.go [hunk 6]
```diff
@@ -644,7 +642,7 @@ func (proof *ChangeProof) Empty() bool {
 // If [start] is nil, there is no lower bound on acceptable keys.
 // If [end] is nil, there is no upper bound on acceptable keys.
 // If [kvs] is empty, returns nil.
-func verifyKeyChanges(kvs []KeyChange, start, end []byte) error {
+func verifyKeyChanges(kvs []KeyChange, start []byte, end Maybe[[]byte]) error {
```
Comment (darioush) [inline x/merkledb/proof.go:645]: could we update the comment for this fn to refer to Nothing instead of nil as appropriate?

File: x/merkledb/db.go [hunk 7]
```diff
@@ -83,7 +83,7 @@ type ChangeProofer interface {
 		ctx context.Context,
 		proof *ChangeProof,
 		start []byte,
-		end []byte,
```
Comment (darioush) [inline x/merkledb/db.go:86]: could we update the comment for this fn to refer to Nothing as appropriate?

File: x/sync/g_db/db_server.go [hunk 8]
```diff
@@ -82,7 +82,7 @@ func (s *DBServer) VerifyChangeProof(
 
 	// TODO there's probably a better way to do this.
 	var errString string
-	if err := s.db.VerifyChangeProof(ctx, &proof, req.StartKey, req.EndKey, rootID); err != nil {
+	if err := s.db.VerifyChangeProof(ctx, &proof, req.StartKey, merkledb.Some(req.EndKey), rootID); err != nil {
```
[the remark this answers is not in the document]
Comment (dboehm-avalabs) [reply x/sync/g_db/db_server.go:85]: This just hasn't been updated yet to support Maybes yet.

File: x/sync/client.go [hunk 9]
```diff
@@ -111,7 +111,12 @@ func (c *client) GetChangeProof(ctx context.Context, req *pb.SyncGetChangeProofR
 			return nil, err
 		}
 
-		if err := db.VerifyChangeProof(ctx, &changeProof, req.StartKey, req.EndKey, endRoot); err != nil {
+		endKey := merkledb.Nothing[[]byte]()
+		if req.EndKey != nil && !req.EndKey.IsNothing {
```
[the remark this answers is not in the document]
Comment (darioush) [reply x/sync/client.go:115]: We can keep the existing semantics for now and possibly avoid the IsNothing bool field in a separate PR

File: x/sync/manager.go [hunk 10]
```diff
@@ -677,20 +677,25 @@ func (m *Manager) enqueueWork(work *workItem) {
...
 	if length == 0 {
-		return []byte{127}
+		if end.IsNothing() {
+			return []byte{127}
+		} else if len(end.Value()) == 0 {
+			return nil
+		}
 	}
 
 	// This check deals with cases where the end has a 255(or is nil which is treated as all 255s) and the start key ends 255.
```
Comment (darioush) [inline x/sync/manager.go:695]: should update this comment

File: x/sync/manager.go [hunk 11]
```diff
@@ -695,7 +696,7 @@ func midPoint(start []byte, end merkledb.Maybe[[]byte]) []byte {
...
 	// For example, midPoint([255], nil) should be [255, 127], not [255].
 	// The result needs the extra byte added on to the end to deal with the fact that the naive midpoint between 255 and 255 would be 255
-	if (len(start) > 0 && start[len(start)-1] == 255) && (end.IsNothing() || end.Value()[len(end.Value())-1] == 255) {
+	if (len(start) > 0 && start[len(start)-1] == 255) && (end.IsNothing() || len(end.Value()) == 0 || end.Value()[len(end.Value())-1] == 255) {
```
Comment (darioush) [inline x/sync/manager.go]: Should this be
```suggestion
	if (len(start) > 0 && start[len(start)-1] == 255) && (end.IsNothing() || len(end.Value()) > 0 && end.Value()[len(end.Value())-1] == 255) {
```

File: x/sync/manager.go [hunk 12]
```diff
@@ -623,7 +629,7 @@ func (m *Manager) completeWorkItem(ctx context.Context, work *workItem, largestH
 	// completed the range [work.start, lastKey], log and record in the completed work heap
 	m.config.Log.Info("completed range",
 		zap.Binary("start", work.start),
-		zap.Binary("end", largestHandledKey),
+		zap.Binary("end", largestHandledKey.Value()),
```
[the remark this answers is not in the document]
Comment (danlaine) [reply x/sync/manager.go:632]: +1 for `String()`

File: x/merkledb/proof.go [hunk 13]
```diff
@@ -352,32 +353,24 @@ func (proof *RangeProof) Verify(
...
+		largestPath,
 	); err != nil {
 		return err
 	}
 	if err := addPathInfo(
 		view,
 		proof.EndProof,
-		insertChildrenLessThan,
-		insertChildrenGreaterThan,
+		Some(smallestPath),
```
Comment (danlaine) [inline x/merkledb/proof.go:372]: Why do we now always pass in `Some` instead of passing in `Nothing` when `len(smallestPath) == 0`?

File: x/sync/manager.go [hunk 14]
```diff
@@ -584,11 +590,11 @@ func (m *Manager) setError(err error) {
 
 // Mark the range [start, end] as synced up to [rootID].
 // Assumes [m.workLock] is not held.
-func (m *Manager) completeWorkItem(ctx context.Context, work *workItem, largestHandledKey []byte, rootID ids.ID, proofOfLargestKey []merkledb.ProofNode) {
+func (m *Manager) completeWorkItem(ctx context.Context, work *workItem, largestHandledKey merkledb.Maybe[[]byte], rootID ids.ID, proofOfLargestKey []merkledb.ProofNode) {
 	// if the last key is equal to the end, then the full range is completed
-	if !bytes.Equal(largestHandledKey, work.end) {
+	if work.end.IsNothing() || !bytes.Equal(largestHandledKey.Value(), work.end.Value()) {
```
Comment (danlaine) [inline x/sync/manager.go:595]: Is this correct? We're checking here whether we're done fetching the range in `work`, and if there's no upper bound on `work` (i.e. `work.end.IsNothing()`) then we always treat it as though we're not done fetching the range. Don't we also need to look at `largestHandledKey`? See [here](https://github.com/ava-labs/avalanchego/pull/1657/files#diff-69e61c5702db139bb75c3063cfee14dd84f123d2febf213b2a8767f368ff1664R617)

File: x/merkledb/proof.go (hunk 13 above)
Comment (dboehm-avalabs) [reply x/merkledb/proof.go:372]: functionally makes no difference, since nil is the smallest possible key, it is equivalent to no bounds.

File: x/merkledb/proof.go [hunk 15]
```diff
@@ -352,32 +353,24 @@ func (proof *RangeProof) Verify(
...
-	insertChildrenGreaterThan := Nothing[path]()
-	if len(largestPath) > 0 {
-		insertChildrenGreaterThan = Some(largestPath)
-	}
 	if err := addPathInfo(
 		view,
 		proof.StartProof,
-		insertChildrenLessThan,
-		insertChildrenGreaterThan,
+		Some(smallestPath),
```
[the remark this answers is not in the document]
Comment (dboehm-avalabs) [reply x/merkledb/proof.go:364]: functionally makes no difference, since nil is the smallest possible key, it is equivalent to no bounds.

File: x/sync/manager.go (hunk 14 above)
Comment (dboehm-avalabs) [reply x/sync/manager.go:595]: I'm not sure I understand.  Since end is unbound, largestHandledKey will always be unequal which means we won't know if the range is complete just via comparison.  We definitely need to use the findnextkey logic to determine range completion.

Comment (danlaine) [reply x/sync/manager.go:595]: > Since end is unbound, largestHandledKey will always be unequal

I don't think this is true, actually. It's possible that `work.end.IsNothing() && largestHandledKey.IsNothing()`. Namely, if we request a range/change proof with no upper bound and receuve a range/change proof with no key-values.

That said, I think it might be a bit more readable to change this to:

```go
if !merkledb.MaybeBytesEquals(largestHandledKey, work.end) {
```

as in the code linked above. I just think it's easier to reason about this with "the largest key we handled wasn't equal to the one we wanted", which is what the code above basically says, rather than:

```
if work.end.IsNothing() || !bytes.Equal(largestHandledKey.Value(), work.end.Value()) {
```

which I think is somewhat less clear.

File: x/merkledb/db.go [hunk 16]
```diff
@@ -83,7 +83,7 @@ type ChangeProofer interface {
...
+		end Maybe[[]byte],
```
Comment (danlaine) [inline x/merkledb/db.go:86]: Should we also use `end Maybe[[]byte]` in `GetChangeProof` above and `GetRangeProofAtRoot` below for consistency?

**Review by danlaine — APPROVED** [review]

File: x/merkledb/db.go (hunk 16 above)
Comment (dboehm-avalabs) [reply x/merkledb/db.go:86]: Part 2

**Review by darioush — APPROVED** [review]

**Review by danlaine — APPROVED** [review]
